### ISSUE TITLE
Test-cases using serenity.batch commands are not being split up

### ISSUE THREAD

**Issue by timothyjmtan** [serenity-bdd/serenity-core#2199]
Hello,

I am having some difficulties in running the JUnit test-cases in batches via the serenity.batch.count commands documented here:

[Serenity Parallel Batches](https://serenity-bdd.github.io/theserenitybook/latest/serenity-parallel-batches.html)

My testing stack is as follows:
Jenkins -> Openshift pod (slave) -> AWS browser stack

Upon Jenkins job trigger, an openshift pod will spin up dynamically, and coordinate with the AWS stack to run the tests.

**Expectation**
By using serenity.batch.count=3, there should be 3 openshift pods spun up, and each will handle a test-suite corresponding to its batch number. 

**Observation**
3 openshift pods spin up, and from console logs it is running the correct mvn command with the correct batch numbers:

Pod #1 - `mvn clean verify -Dserenity.batch.number=1 -Dserenity.batch.count=3` --> runs 0 test-cases
Pod #2 - `mvn clean verify -Dserenity.batch.number=2 -Dserenity.batch.count=3` --> runs 0 test-cases
Pod #3 - `mvn clean verify -Dserenity.batch.number=3 -Dserenity.batch.count=3` --> runs 200 test-cases

However, 2 pods will execute 0 test-cases, and 1 pod will run the entire test-suite. This suggests that the splitting is not being handled properly somewhere.

No difference with the inclusion of the `serenity.batch.strategy` parameter.

It has the same behaviour when executing on plain windows machines as well.

I don't reckon I need to set the following in `maven-failsafe-plugin` in pom.xml, as its for subdivisions within a test machine.

```
        <parallel>classes</parallel>
        <threadCount>${parallel.tests}</threadCount>
        <forkCount>${parallel.tests}</forkCount>
```

Is there anywhere that I can take a look at to troubleshoot?

**Comment by wakaleo** [comment]
This would be a question for @nbarrett.

**Comment by nbarrett** [comment]
Okay @timothyjmtan, here are some thoughts...

- When it comes to diagnosing a problem, my strong recommendation would be to keep the complexity to a minimum, so I would suggest removing the Openshift pod (slave) and AWS browser stack from the equation, until you understand and get the batching working properly. In order to achieve this, my suggestion would be to try and run these batches locally on your dev machine first to work out what's going on.
- Fortunately, Serenity has a most excellent "dry run" mode which can be activated using `-Dserenity.dry.run=true`. This very quickly runs all tests but bypasses all webdriver navigation, so your suite can be run in just a few seconds. 
- I'm assuming that you have used the Screenplay pattern, but in dry run mode, only the "outer" `Performable`s will be executed. This is also a very quick way of working out whether your entire test suite can even be run properly. I'm not sure whether you've had full, clean runs from your suite yet, so thought I'd mention this. Dry run can highlight some design problems in your test setup or `Performable`s that will prompt some refactoring or redesign (e.g, class initialisation problems or a host of other things that need to be fixed during the suite refinement stage).
- Moving onto the maven configuration, I would recommend that you have something along these lines to ensure that tests are ordered correctly. I've helped other people in the past with batching problems because they didn't have `runOrder` configured properly  (please excuse my out of date versions, I've not used maven for about 4 years):
```
    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>2.19.1</version>
                <configuration>
                    <runOrder>alphabetical</runOrder>
                    <systemPropertyVariables>
                    </systemPropertyVariables>
                    <skipTests>${skip.surefire.tests}</skipTests>
                </configuration>
            </plugin>
            <plugin>
                <artifactId>maven-failsafe-plugin</artifactId>
                <version>2.19.1</version>
                <configuration>
                    <runOrder>alphabetical</runOrder>
                    <includes>
                        <include>com.blah.blah/**.java</include>
                    </includes>
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>integration-test</goal>
                            <goal>verify</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>net.serenity-bdd.maven.plugins</groupId>
                <artifactId>serenity-maven-plugin</artifactId>
                <version>${serenity.maven.plugin.version}</version>
                <dependencies>
                    ...
                </dependencies>
            </plugin>
        </plugins>
    </build>
```
-  Another thing that I've found is useful with Junit test suites that are batched is to tag the tests so that you can see in the html report how the tests are balanced across the slices. Here's something you could add to your Test runner(s) to help provide insight:

```
@After
    public void tearDown() {
        tagTestWithSlice();
    }

    public static void tagTestWithSlice() {
        EnvironmentVariables environmentVariables = Injectors.getInjector().getInstance(EnvironmentVariables.class);
        TestTag testTag = withName(String.format("Slice %02d", valueOf(environmentVariables.getProperty(SERENITY_BATCH_NUMBER, "1")))).andType("Test Slice");
        StepEventBus.getEventBus().addTagsToCurrentTest(newArrayList(testTag));
    }
```

Hope this helps!

**Comment by timothyjmtan** [comment]
Hello @nbarrett, thanks for the in-depth suggestions!

Just some updates:

- I have significantly reduced the overall complexity by running the command `mvn clean verify -Dserenity.batch.count=4 -Dserenity.batch.number=1` on my local PC. My `test.folder` is pointing to a package containing only 4 tests (written in Screenplay pattern), so the expectation here is that it should only attempt to execute 1 test-case.

- I have introduced `-Dserenity.dry.run=true`, and have checked that all 4 test-cases are running w/o issues using a vanilla `mvn clean verify`.

- With the only change being the inclusion of the two additional parameters `mvn clean verify -Dserenity.batch.count=4 -Dserenity.batch.number=1`, I am getting 0 tests being executed.

- No difference between `-Dserenity.batch.count=4` and `-Dserenity.batch.size=4`. I read the source code and saw that the former is actually deprecated.

- I have also tried incrementing the `-Dserenity.batch.number=X` to see if there were any differences. All yielded 0 tests, except for `serenity.batch.number=4` which executed all 4 test-cases.

- Have incrementally introduced the suggestions in the pom.xml configurations, but I do not see any difference in behaviour (i.e. 0 tests executed).

So it seems to be that there is something iffy with the batch allocation of the tests. Are there are any lower-level logs which I can access to see where the issue lies in? From where I currently stand, I'm not entirely sure if it is a maven or serenity issue...

**Comment by nbarrett** [comment]
Hi again @timothyjmtan - well it does seem like you are being very methodical in your approach and doing all the right things. I now recall that I did have a very similar problem to you and raised #736 but ended up resolving it by fixing the maven plugin configuration as described in [this note](https://github.com/serenity-bdd/serenity-core/issues/736#issuecomment-426165743). Around this time, I started work on a batch refactoring PR which included:
- addition of logging in BatchManagerProvider to help people diagnose why tests are being included/excluded
- updated tests to use more modern Java constructs
- refactored THUCYDIDES properties to SERENITY equivalents
- fixed Demo projects and updated dependencies to enable them to be run to demonstrate the batching

However due to resolving my problem, there was no need at the time to make fixes changes to serenity. I think it might be a good idea if I bring this up to date and then submit it, then hopefully the logging will be available in the core product. What does that sound like?

**Comment by timothyjmtan** [comment]
Hello @nbarrett , yeah I suppose that would be useful for anyone who is running into issues using the batching functionality 👍 

Just curious, is it a must to include the `-Dparallel="classes" -Dparallel.count=4 -DuseUnlimitedThreads=true` parameters for basic batching functionality to work? From my understanding these are just to enable concurrency testing within the subset of a batch, so it's not required per se.

**Comment by nbarrett** [comment]
I've never personally used any of those parameters, it should work with just the two we've already discussed. Okay will try and get that PR submitted later on 👍

**Comment by nbarrett** [comment]
I've got one last thought.... when you mention the 4 tests in your diagnosis, how are they allocated? For instance are there 4  `@Test` methods or 4 classes? I would suggest splitting them up into separate classes as I think the batching doesn't go down to the test method level.

**Comment by timothyjmtan** [comment]
Hello @nbarrett , it's 4 classes, each containing 1 @Test annotation. As such I believe the batching should work... Hope that clarifies!

**Comment by timothyjmtan** [comment]
Howdy @nbarrett, after digging through the code, I think I have found the issue.

Based on my understanding of how the batching works within the BatchManager:
- As each test-case is 'handled' by the framework, [it will be registered](https://github.com/serenity-bdd/serenity-core/blob/f8ef5b90b1c6ef04bd876a6a59fd996bcc2bb228/serenity-model/src/main/java/net/thucydides/core/batches/SystemVariableBasedBatchManager.java#L64) and the atomic integer `testCaseCount` will be incremented by 1.
- After this increment, the method `shouldExecuteThisTest()` will be executed.
- Based on the current value of `testCaseCount.get()`, the method will return a `true` or `false` depending on the [outcome of the logic check](https://github.com/serenity-bdd/serenity-core/blob/f8ef5b90b1c6ef04bd876a6a59fd996bcc2bb228/serenity-model/src/main/java/net/thucydides/core/batches/SystemVariableBasedBatchManager.java#L71)

What I've noticed is that in my code, before any test-case is actually 'executed', the batch manager is parsing through every single test-case, and doing the `testCaseCount` increments. As such, when the test-case is actually being executed, the `testCaseCount` is already at the max value, i.e. the total number of test-cases, and as such the logic breaks.

With an example, my code is currently giving the following values:

Number of test-cases: 4
serenity.batch.number=1
serenity.batch.count=4

TestCaseA.java --> expected: `testCaseCount=1` actual: `testCaseCount=4`  
TestCaseB.java --> expected: `testCaseCount=2` actual: `testCaseCount=4`
TestCaseC.java --> expected: `testCaseCount=3` actual: `testCaseCount=4`
TestCaseD.java --> expected: `testCaseCount=4` actual: `testCaseCount=4`

As such, batches 1-3 will always return 0 TCs to execute, and `serenity.batch.number=4` will execute everything.

It is worth noting that I am actually using an older version of Serenity, i.e. `1.9.36`. Comparing it to the latest version in github, I do not see any changes in logic, but perhaps there was one in `serenity-core`?

**Comment by nbarrett** [comment]
Hi @timothyjmtan - great diagnostics and very similar to what I did when I was exploring back here in [#736](https://github.com/serenity-bdd/serenity-core/issues/736#issuecomment-426165743). The [BatchManager](https://github.com/serenity-bdd/serenity-core/blob/master/serenity-model/src/main/java/net/thucydides/core/batches/BatchManager.java#L6) implementations do rely on being executed in the same sequence and in such a way that the atomic integer increments gradually throughout the execution of the suite and is not reset between each test case. For some reason it seems that some people's configuration (yours included) causes these conditions to not occur resulting in incorrect batching behaviour. I did spend some time on the PR last night and cleaned things up quite a but and added logging, but I didn't make any changes to the batching algorithm itself. My preference would be to implement the same approach for batching as has been done for [Cucumber](https://serenity-bdd.github.io/theserenitybook/latest/serenity-parallel-batches.html#_cucumber_specific_configuration), namely using MultiRunTestStatistics, as this works really well. The challenge I've yet to resolve is that this algorithm needs to build a list of all tests in the suite before any tests are run. I just need to work out how to hook this point in the lifecycle reliably. I'll do some more work on this later on and hopefully get that in the PR as well.

**Comment by timothyjmtan** [comment]
Hey @nbarrett , just to give another update. We've managed to pinpoint and resolve the issue.

We're using a parent-child pom model in our project, and it turns out that the parent pom's `maven-failsafe` plugin contained the following settings:
```
<parallel>classes</parallel>
<threadCountClasses>1</threadCountClasses>
<forkCount>${parallel.count}</forkCount>
```
Commenting these out will allow the batching to be done, so problem solved!

However this gives us another issue where we are now not able to execute the tests within each batch in parallel (via forking). Is this a functionality that can be taken a look at and fixed?

**Comment by wakaleo** [comment]
Why not check the code and see what it is doing? If it uses thread-local counters then parallel execution for a given batch will not work.

**Comment by dicksonprof** [comment]
@wakaleo  Does not seems like forking or parallel threading runs would work.
Is there any way to replication the CucumberSlicerRunner login into the Junit.
Some ideas could be writting a maven-plugin that runs at pre-integration-test to do the splititing and writting to a file.
Maybe using Reflection to scan @Test and @RunWith(SerenityRunner), SerenityParameterizedRunner

After that the BatchManager reads this file during integrationTest phases. This will probably be more deterministic but the logic and computation may take some time and i think it should be able to support forking/threading

**Comment by nbarrett** [comment]
Yes, that kind of thing was what I had in mind. I'll try and finish my PR this weekend.

**Comment by thePantz** [comment]
Hello, I've just come across this problem. Has this been solved?

**Comment by nbarrett** [comment]
Sorry @thePantz -  I now recall I promised to complete my PR but didn't find time yet. Will try and do soon!

**Comment by azmo-rinsler** [comment]
Has there been any movement on this issue?  I've been pulling my hair out trying to get batching to work, to no avail...

**Comment by azmo-rinsler** [comment]
Has there been any movement on this?  I'm testing a Java Swing application, and the driver I have to use when interacting with it does not work well with threads... batching previously allowed me to get around this by running each of them in their own Runtime environment, but I can't seem to find any combination of settings that works?

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>serenity.cucumber.disaster</groupId>
    <artifactId>serenity-cucumber-disaster</artifactId>
    <version>1.0</version>

    <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <encoding>UTF-8</encoding>

        <serenity.version>2.4.34</serenity.version>
        <serenity.maven.version>2.4.34</serenity.maven.version>
        <serenity.cucumber.version>2.4.34</serenity.cucumber.version>

        <tags/>
        <webdriver.base.url/>

        <!-- This placeholder should be set at runtime using the -Dbatches arg -->
        <batches>1</batches>
    </properties>

    <dependencies>
        <dependency>
            <groupId>net.serenity-bdd</groupId>
            <artifactId>serenity-core</artifactId>
            <version>${serenity.version}</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>net.serenity-bdd</groupId>
            <artifactId>serenity-cucumber6</artifactId>
            <version>${serenity.cucumber.version}</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>net.serenity-bdd</groupId>
            <artifactId>serenity-screenplay</artifactId>
            <version>${serenity.version}</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>net.serenity-bdd</groupId>
            <artifactId>serenity-screenplay-webdriver</artifactId>
            <version>${serenity.version}</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>net.serenity-bdd</groupId>
            <artifactId>serenity-ensure</artifactId>
            <version>${serenity.version}</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>junit</groupId>
            <artifactId>junit</artifactId>
            <version>4.13.1</version>
            <scope>test</scope>
        </dependency>

        <dependency>
            <groupId>ch.qos.logback</groupId>
            <artifactId>logback-classic</artifactId>
            <version>1.0.13</version>
        </dependency>

        <dependency>
            <groupId>org.assertj</groupId>
            <artifactId>assertj-core</artifactId>
            <version>3.6.2</version>
            <scope>test</scope>
        </dependency>
    </dependencies>
    <build>
        <plugins>
            <!-- Kotlin Stuff -->
            <plugin>
                <groupId>org.jetbrains.kotlin</groupId>
                <artifactId>kotlin-maven-plugin</artifactId>
                <!--suppress DifferentKotlinMavenVersion -->
                <version>1.5.0</version>
                <executions>
                    <execution>
                        <id>compile</id>
                        <phase>none</phase>
                        <goals>
                            <goal>compile</goal>
                        </goals>
                        <configuration>
                            <sourceDirs>
                                <sourceDir>${project.basedir}/src/main/kotlin</sourceDir>
                                <sourceDir>${project.basedir}/src/main/java</sourceDir>
                            </sourceDirs>
                        </configuration>
                    </execution>
                    <execution>
                        <id>test-compile</id>
                        <goals>
                            <goal>test-compile</goal>
                        </goals>
                        <configuration>
                            <sourceDirs>
                                <sourceDir>${project.basedir}/src/test/kotlin</sourceDir>
                                <sourceDir>${project.basedir}/src/test/java</sourceDir>
                            </sourceDirs>
                        </configuration>
                    </execution>
                </executions>
            </plugin>

            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>3.0.0-M5</version>
                <configuration>
<!--    These do not appear to have any effect ?
                    <runOrder>alphabetical</runOrder>
                    <skip>true</skip>
-->
                </configuration>
            </plugin>
            <plugin>
                <artifactId>maven-failsafe-plugin</artifactId>
                <version>3.0.0-M5</version>
                <configuration>

                    <includes>
                        <include>**/*TestRunner.java</include>
                    </includes>

                    <systemPropertyVariables>
                        <webdriver.base.url>${webdriver.base.url}</webdriver.base.url>
<!--    Enabling these immediately breaks filtering - generally resulting in NO tests being run
                        <serenity.batch.count>0${batches}</serenity.batch.count>
                        <serenity.batch.number>0${surefire.forkNumber}</serenity.batch.number>
-->
<!--    These do not appear to have any effect ?
                        <serenity.batch.strategy>DIVIDE_BY_TEST_COUNT</serenity.batch.strategy>
                        <serenity.test.statistics.dir>/aggregated</serenity.test.statistics.dir>
-->
<!--    Behavior is not noticeably different from the <serenity.batch...> variants
                        <serenity.fork.count>0${batches}</serenity.fork.count>
                        <serenity.fork.number>0${surefire.forkNumber}</serenity.fork.number>
-->
                    </systemPropertyVariables>

<!--    Threads seem to work well as long as NONE of the serenity batch/fork properties are enabled, and there is
        only 1 test 'Runner' class defined (multiple runners result in each of them running all tests)
                    <parallel>classes</parallel>
                    <threadCount>${batches}</threadCount>
-->
<!--    These do not appear to have any effect ?
                    <runOrder>alphabetical</runOrder>
                    
                    <forkCount>${batches}</forkCount>
                    <reuseForks>false</reuseForks>
                    
                    <useManifestOnlyJar>false</useManifestOnlyJar>
                    <useManifestOnlyJar>true</useManifestOnlyJar>
-->
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>integration-test</goal>
                            <goal>verify</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.8.1</version>
                <executions>
                    <execution>
                        <id>default-compile</id>
                        <phase>none</phase>
                    </execution>
                    <execution>
                        <id>default-testCompile</id>
                        <phase>none</phase>
                    </execution>
                    <execution>
                        <id>java-compile</id>
                        <phase>compile</phase>
                        <goals>
                            <goal>compile</goal>
                        </goals>
                    </execution>
                    <execution>
                        <id>java-test-compile</id>
                        <phase>test-compile</phase>
                        <goals>
                            <goal>testCompile</goal>
                        </goals>
                    </execution>
                </executions>
                <configuration>
                    <source>1.8</source>
                    <target>1.8</target>
                </configuration>
            </plugin>
            <plugin>
                <groupId>net.serenity-bdd.maven.plugins</groupId>
                <artifactId>serenity-maven-plugin</artifactId>
                <version>${serenity.maven.version}</version>
                <configuration>
                    <tags>${tags}</tags>
                </configuration>
                <dependencies>
                    <dependency>
                        <groupId>net.serenity-bdd</groupId>
                        <artifactId>serenity-core</artifactId>
                        <version>${serenity.version}</version>
                    </dependency>
                </dependencies>
                <executions>
                    <execution>
                        <id>serenity-reports</id>
                        <phase>post-integration-test</phase>
                        <goals>
                            <goal>aggregate</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>
</project>
```
